src/chainstate/burn/operations/leader_block_commit.rs

```diff
@@ -17,36 +17,34 @@
  along with Blockstack. If not, see <http://www.gnu.org/licenses/>.
 */
 
-use chainstate::operations::{BlockstackOperation, BlockstackOperationType};
-use chainstate::operations::Error as op_error;
-use chainstate::{ConsensusHash, BlockHeaderHash, VRFSeed};
+use std::marker::PhantomData;
 
-use chainstate::db::burndb::BurnDB;
+use chainstate::burn::operations::BlockstackOperation;
+use chainstate::burn::operations::Error as op_error;
+use chainstate::burn::{BlockHeaderHash, VRFSeed};
+
+use chainstate::burn::db::burndb::BurnDB;
+use chainstate::burn::db::DBConn;
 
 use burnchains::{BurnchainTransaction, BurnchainTxInput, PublicKey};
-use burnchains::bitcoin::keys::BitcoinPublicKey;
-use burnchains::bitcoin::indexer::BitcoinNetworkType;
-use burnchains::bitcoin::address::{BitcoinAddressType, BitcoinAddress};
 use burnchains::Txid;
-use burnchains::Hash160;
 use burnchains::Address;
+use burnchains::BurnchainHeaderHash;
 
 use util::hash::hex_bytes;
-
-use ed25519_dalek::PublicKey as VRFPublicKey;
-
-use crypto::sha2::Sha256;
+use util::log;
 
 pub const OPCODE: u8 = '[' as u8;
 
 #[derive(Debug, PartialEq, Clone)]
-pub struct LeaderBlockCommitOp<K: PublicKey> {
-    pub block_header_hash: BlockHeaderHash, // hash of block header (double-sha256)
+pub struct LeaderBlockCommitOp<A, K> {
```

Comment:
Why are Address and Public Key generics?

Reply:
The public key and address types are specific to the burn chain on top of which this algorithm operates.  We'll be adding support for both Bitcoin as a burn chain, as well as Stacks as a burn chain.  They'll have different Address and Public Key types.

I can make the address and public key types into associated types instead of type parameters if it'll make the code easier to parse.

Reply:
I was just wondering if we were going to have multiple implementations of those traits. But as far as the Stacks burn chain goes, are we not using the same elliptic curve as Bitcoin for keys? If so, why would they be different address or public key implementations?

Reply:
I *think* we are using the same elliptic curve, but nevertheless I don't want to accidentally "cross wires" in the code.  Giving them different types ensures that we don't accidentally pass one chain's public key into code that only operates on another chain's data.

Reply:
I guess what I'm really after is a "burn chain type", which contains associated "public key" and "address" types.  Then, we'd say "this is a LeaderBlockCommitOp for the Stacks chain" or "this is a LeaderBlockCommitOp for the Bitcoin chain" instead of what we have now.

Reply:
Ah, I see. That makes sense.